sporkfish/searcher/minimax.py

```diff
@@ -0,0 +1,253 @@
+import copy
+import logging
+import os
+import time
+from abc import ABC, abstractmethod
+from typing import Any, Callable, List, Optional, Tuple
+
+import chess
+import stopit
+from pathos.multiprocessing import ProcessPool
+
+from ..board.board import Board
+from ..evaluator import Evaluator
+from ..statistics import Statistics
+from ..transposition_table import TranspositionTable
+from ..zobrist_hasher import ZobristHasher
+from .move_ordering import MoveOrder, MvvLvaHeuristic
+from .searcher import Searcher
+from .searcher_config import MoveOrdering, SearcherConfig, SearchMode
+
+_dict: dict = dict()
+
+
+class MiniMaxVariants(Searcher):
+    """
+    Abstract base class for minimax like searchers
+    """
+
+    @abstractmethod
+    def _searcher(
+        self, board_to_search: Board, depth: int, alpha: float, beta: float
+    ) -> Tuple[float, chess.Move]:
+        pass
+
+    def __init__(
+        self,
+        evaluator: Evaluator,
+        move_order: MoveOrder,
+        config: SearcherConfig = SearcherConfig(),
+    ) -> None:
+        super().__init__(evaluator, config)
+
+        if self._config.enable_transposition_table:
+            self._zobrist_hash = ZobristHasher()
+            self._transposition_table = TranspositionTable(_dict)
+            logging.info("Enabled transposition table in search.")
+        else:
+            logging.info("Disabled transposition table in search.")
+
+        self._move_order = move_order
+
+    def _ordered_moves(self, board: Board, legal_moves: Any) -> Any:
+        # order moves from best to worse
+        return sorted(
+            legal_moves,
+            key=lambda move: (self._move_order.evaluate(board, move),),
+            reverse=True,
+        )
+
+    def _aspiration_windows_search(
+        self,
+        search_func: Callable[[Board, int, float, float], Tuple[float, chess.Move]],
```

Comment:
I think this can be removed, directly calling your currently called `_searcher` - easier on the eyes

Reply:
At the moment, the `MVV_LVA` matrix is hard coded in the function:

```
# make this configurable in future (less error prone)
        # Columns: attacker P, N, B, R, Q, K
        MVV_LVA = [
            [15, 14, 13, 12, 11, 10],  # victim P
            [25, 24, 23, 22, 21, 10],  # victim N
            [35, 34, 33, 32, 31, 30],  # victim B
            [45, 44, 43, 42, 41, 40],  # victim R
            [55, 54, 53, 52, 51, 50],  # victim Q
            [0, 0, 0, 0, 0, 0],  # victim K
        ]
```
There are a couple of reasons I don't quite like hard coded numbers (in general, not just about this matrix):

- Seems a bit like a black box, no one really understands the behaviour and heuristics of these numbers - and it is hard to test
- Makes it a bit hard to unit test and maintain:
	- The way the MVV LVA unit tests works is by setting up simple board that allows for a small set of capturing legal moves, and try to replicate the numbers in the test. Since this matrix is not accessible outside of the scope of the function, I had to hard code the numbers to match the values. 
	- If (or when) someone does change something in this matrix, either the test will break, and the developer would need to fix the test manually, or worse, the change is not tested.

Hence, my suggestion is either (a) make this matrix configurable, or (b) make this matrix a member (@property) within the class, so at least it is accessible by clients.

Of course, this is not high priority, and suspect no plans to play around with the numbers in this matrix. 

LMKYT?

Reply:
Good catch! - fixed!